[PS4,PS5][Driver] Detangle --sysroot and -isysroot

playstation-edd opened the pull request:
The following discrepancies concerning `-isysroot` and `--sysroot` motivated this change:

- The SDK directory can be specified via `-isysroot`, but `--sysroot` has no influence over this. Yet, we check for the presence of either switch to determine whether we ought to warn about a missing SDK *headers*.

- The presence of `-isysroot` is ignored when deciding whether to warn about missing SDK *libraries*, depsite it being the only switch capable of specifying a non-default SDK location.

- The `--sysroot`s passed to the PlayStation linkers by the driver are unrelated to the SDK directory resolved in the PS4PS5Base constructor.

(We also ought to pass a sysroot-relative library search path to the linker via `-L=/target/lib`, but that's for another PR).

So we're half way between two worlds, currently:

- World 1: `-isysroot` and `--sysroot=` mean the same thing and each may override the SDK directory, from which both header and library search paths are derived.
- World 2: `-isysroot` influences header search paths and `--sysroot=` influences library search paths. Each independently defaults to the SDK directory.

This change commits to world 2, which seems to offer more flexibility for the user and better adheres to the principle of least surprise for those familiar with these options outside of a PlayStation development environment.

The test updates to ps{4,5}-sdk-root.c were of the scale of a rewrite so I also took the opportunity to clarify the purpose of each part, eliminate some redundancy and add some missing coverage.

SIE tracker: TOOLCHAIN-16704

playstation-edd committed:
[PS4,PS5][Driver] Detangle --sysroot and -isysroot

The following discrepancies concerning `-isysroot` and `--sysroot`
motivated this change:

- The SDK directory can be specified via `-isysroot`, but `--sysroot`
  has no influence over this. Yet, we check for the presence of either
  switch to determine whether we ought to warn about a missing SDK
  *headers*.

- The presence of `-isysroot` is ignored when deciding whether to warn
  about missing SDK *libraries*, depsite it being the only switch
  capable of specifying a non-default SDK location.

- The `--sysroot`s passed to the PlayStation linkers by the driver are
  unrelated to the SDK directory resolved in the PS4PS5Base constructor.

(We also ought to pass a sysroot-relative library search path to the
linker via `-L=/target/lib`, but that's for another PR).

So we're half way between two worlds, currently:

World 1: `-isysroot` and `--sysroot=` mean the same thing and each may
         override the SDK directory, from which both header and library
         search paths are derived.
World 2: `-isysroot` influences header search paths and `--sysroot=`
         influences library search paths. Each independently defaults
         to the SDK directory.

This change commits to world 2, which seems to offer more flexibility
for the user and better adheres to the principle of least surprise for
those familiar with these options outside of a PlayStation development
environment.

The test updates to ps{4,5}-sdk-root.c were of the scale of a rewrite so
I also took the opportunity to clarify the purpose of each part,
eliminate some redundancy and add some missing coverage.

SIE tracker: TOOLCHAIN-16704

pogo59 reviewed (commented):
According to [gcc docs](https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html) `--sysroot` applies to headers and libraries, and `-isysroot` overrides it for headers only. This is different from the "independence" described for World 2. That is, if we have only `--sysroot` it should apply to both.

pogo59 commented:
FTR I have no good explanation for why the PS4 handling is in the state it's in--thanks for sorting this!

playstation-edd committed:
Allow -isysroot to override --sysroot for header search

A summary of how search roots are now established:

* The SDK root is sought in the environment, with a fallback to the
  driver's location.
* The library and header search paths default to being rooted at the SDK
  root, but can be overridden with --sysroot.
* The header search path root can be further overridden with -isysroot,
  which will also override any header root due to --sysroot.

playstation-edd commented:
> According to [gcc docs](https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html) `--sysroot` applies to headers and libraries, and `-isysroot` overrides it for headers only. This is different from the "independence" described for World 2. That is, if we have only `--sysroot` it should apply to both.

I'm not sure what resource is (mis)read to come to the conclusion that world 2 was normal. Thanks for checking!

I have pushed an update which has `--sysroot` apply to both header and libraries, unless both `--sysroot` and `-isysroot` are supplied, win which case the former applies to libraries and the latter to headers.

pogo59 commented on clang/lib/Driver/ToolChains/PS4CPU.cpp:
```diff
@@ -323,46 +323,63 @@ toolchains::PS4PS5Base::PS4PS5Base(const Driver &D, const llvm::Triple &Triple,
+    SmallString<128> Dir(SDKLibraryRootDir);
+    llvm::sys::path::append(Dir, "target/lib");
+    if (CheckSDKPartExists(Dir, "system libraries"))
+      getFilePaths().push_back(std::string(Dir));
```
This `push_back` doesn't happen if there's a custom sysroot? Offhand I don't recall what the effect of FilePaths is, but I think that's not what we want.

playstation-edd replied:
There's an early `return` in a similar place/situation, prior to this PR. So this was my attempt at NFC.

But I also don't know much about the FilePaths. I'll do some digging to ensure this remains reasonable.

playstation-edd replied:
It's less involved than I was expecting: the contents of the FilePaths are turned into `-L` arguments for the linker.

So I guess the idea is:

* If you're passing `--sysroot`, you want full control. You can pass `<whatever>/target/lib` if you want, but it shouldn't be _assumed_ you want this library path.
* Otherwise, `<SDKROOT>/target/lib` should be added to the linker's search paths. But only if it exists.

I think I buy it.

pogo59 reviewed (approved):
LGTM, but pointed out the need for a test comment update (& replicate to the PS5 equivalent test).

playstation-edd committed:
Clarify descriptions of behaviour in tests.

playstation-edd committed:
Further increase specificity of test descriptions.

playstation-edd commented on clang/test/Driver/ps4-sdk-root.c:
```diff
@@ -1,42 +1,64 @@
+/// If the user takes control of header or library search, the respective
+/// existence check is skipped. User control of header search is assumed if
+/// `--sysroot`, `-isysroot`, `-nostdinc` or `-nostdlibinc` is supplied. User
+/// control of library search is assumed if `--sysroot` is supplied.
```
Fixed - thanks.
